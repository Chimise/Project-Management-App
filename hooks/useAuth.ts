import { useContext } from "react";
import {AuthContext} from "../store/AuthContext";

const useAuth = ()  => {
    const ctx = useContext(AuthContext);
    if(!ctx) {
        throw new Error("Auth context not avialable");
    }

    return ctx;
}

export default useAuth;