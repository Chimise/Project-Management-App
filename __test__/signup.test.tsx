import { getAllByLabelText, prettyDOM, render, user, waitFor, waitForElementToBeRemoved, act } from "../test/test-utils";
import "@testing-library/jest-dom";
import SignUpPage from "../pages/auth/signup";
import "whatwg-fetch";
import server from "../test/server";
import useAuth from '../hooks/useAuth';
import router, {mockUseRouter, mockPush} from '../test/router';

const login = jest.fn();
const logout = jest.fn();

jest.mock('../hooks/useAuth', () => {
    return jest.fn(() => ({
        loginHandler: login,
        token: '',
        logoutHandler: logout,
        isLoading: false
    }));
});

beforeAll(() => {
  server.listen({
    onUnhandledRequest(req) {
      console.log(`Request ${req.url.pathname} ${req.method}`);
    },
  });
});

afterEach(() => {
  server.resetHandlers();
  jest.useRealTimers();
});

beforeEach(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  server.close();
  jest.restoreAllMocks();
});

mockUseRouter();

test("Render the Sign Up page unchanged", () => {
  const { container } = render(<SignUpPage />);
  expect(container.firstChild).toMatchSnapshot();
});

test("Check for error when trying to signup with invalid data", async () => {
  const { getByLabelText, getByRole, findByText } = render(<SignUpPage />);
  const nameInput = getByLabelText("name");
  const emailInput = getByLabelText("email");
  const passwordInput = getByLabelText("password");
  const button = getByRole("button", { name: /sign up/i });

  expect.assertions(2);

  await user.type(nameInput, "chisom");
  await user.type(emailInput, "hello");
  await user.type(passwordInput, "user");

  await user.click(button);

  expect(emailInput).toHaveErrorMessage(/enter a valid email/i);
  expect(passwordInput).toHaveErrorMessage(/your password should be/i);

});

test("Submit an email already in existence and check for an error", async () => {
    const {getByLabelText, getByRole, findByRole} = render(<SignUpPage />)
   await user.type(getByLabelText('name'), 'chisom');
   await user.type(getByLabelText('email'), 'task@task.com');
   await user.type(getByLabelText('password'), 'password');

   await user.click(getByRole('button', {name: /sign up/i}));

   const alert = await findByRole('alert', {description: /email already exist/i});

   expect(alert).toBeInTheDocument();

   act(() => {
    jest.advanceTimersByTime(2500);
   })

   await waitForElementToBeRemoved(alert);
   
})

test('Submit a valid data and check that the user is redirected', async () => {
    const {getByLabelText, getByRole} = render(<SignUpPage />)

   await user.type(getByLabelText('name'), 'chisom');
   await user.type(getByLabelText('email'), 'taskr@taskr.com');
   await user.type(getByLabelText('password'), 'password');

   await user.click(getByRole('button', {name: /sign up/i}));


   await waitFor(() => expect(login).toHaveBeenCalled());
})
