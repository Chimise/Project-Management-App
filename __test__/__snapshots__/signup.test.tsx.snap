// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render the Sign Up page unchanged 1`] = `
<div
  class="md:flex md:flex-row-reverse h-screen"
  style="opacity: 1; transform: none;"
>
  <div
    class="md:flex-1 h-full py-5 md:py-10"
  >
    <div
      class="w-[80%] flex flex-col h-full md:w-[80%] mx-auto"
    >
      <div
        class="shrink-0"
      >
        <a
          href="/"
        >
          <div
            class="flex items-center space-x-2"
          >
            <div
              class="inline-flex items-center justify-center rounded-full p-2 bg-primary"
            >
              <svg
                aria-hidden="true"
                class="w-5 h-5 text-white"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6.429 9.75L2.25 12l4.179 2.25m0-4.5l5.571 3 5.571-3m-11.142 0L2.25 7.5 12 2.25l9.75 5.25-4.179 2.25m0 0L21.75 12l-4.179 2.25m0 0l4.179 2.25L12 21.75 2.25 16.5l4.179-2.25m11.142 0l-5.571 3-5.571-3"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </div>
            <div
              class="text-primary font-bold text-2xl font-sans"
            >
              taskr
            </div>
          </div>
        </a>
      </div>
      <div
        class="flex-1 flex items-center justify-center"
      >
        <div
          class="w-full max-w-sm"
        >
          <h2
            class="text-3xl text-gray-700 text-center md:text-4xl font-semibold"
          >
            Create Account
          </h2>
          <form
            class="space-y-6 my-6"
          >
            <label
              class="block space-y-px group w-full"
            >
              <span
                class="block capitalize group-focus-within:font-medium tracking-wide text-gray-600 font-normal text-xs"
              >
                name
              </span>
              <input
                aria-errormessage="name-error"
                aria-invalid="false"
                autocapitalize="off"
                autocomplete="off"
                autocorrect="off"
                class="block w-full py-0.5 text-sm border-0 px-0 transition-colors duration-100 placeholder:text-sm placeholder-gray-600 focus:ring-0 border-b-2 border-gray-200 focus:outline-none focus:border-slate-700 placeholder-gray-600"
                name="name"
                spellcheck="false"
                type="text"
                value=""
              />
              <span
                class="relative w-full block"
              />
            </label>
            <label
              class="block space-y-px group w-full"
            >
              <span
                class="block capitalize group-focus-within:font-medium tracking-wide text-gray-600 font-normal text-xs"
              >
                email
              </span>
              <input
                aria-errormessage="email-error"
                aria-invalid="false"
                autocapitalize="off"
                autocomplete="off"
                autocorrect="off"
                class="block w-full py-0.5 text-sm border-0 px-0 transition-colors duration-100 placeholder:text-sm placeholder-gray-600 focus:ring-0 border-b-2 border-gray-200 focus:outline-none focus:border-slate-700 placeholder-gray-600"
                name="email"
                spellcheck="false"
                type="email"
                value=""
              />
              <span
                class="relative w-full block"
              />
            </label>
            <label
              class="block space-y-px group w-full"
            >
              <span
                class="block capitalize group-focus-within:font-medium tracking-wide text-gray-600 font-normal text-xs"
              >
                password
              </span>
              <div
                class="w-full relative"
              >
                <input
                  aria-errormessage="password-error"
                  aria-invalid="false"
                  autocapitalize="off"
                  autocomplete="off"
                  autocorrect="off"
                  class="block w-full py-0.5 text-sm border-0 px-0 transition-colors duration-100 placeholder:text-sm placeholder-gray-600 focus:ring-0 border-b-2 border-gray-200 focus:outline-none focus:border-slate-700"
                  name="password"
                  spellcheck="false"
                  type="password"
                  value=""
                />
                <span
                  aria-label="view-password"
                  class="absolute bg-transparent text-gray-400 flex items-center justify-center inset-y-0 right-0 cursor-pointer w-6"
                  role="button"
                >
                  <svg
                    aria-hidden="true"
                    class="w-4 h-4"
                    fill="currentColor"
                    viewBox="0 0 20 20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M10 12.5a2.5 2.5 0 100-5 2.5 2.5 0 000 5z"
                    />
                    <path
                      clip-rule="evenodd"
                      d="M.664 10.59a1.651 1.651 0 010-1.186A10.004 10.004 0 0110 3c4.257 0 7.893 2.66 9.336 6.41.147.381.146.804 0 1.186A10.004 10.004 0 0110 17c-4.257 0-7.893-2.66-9.336-6.41zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </div>
              <span
                class="relative w-full block"
              />
            </label>
            <button
              class="w-full block text-center py-2 bg-black text-white"
              type="submit"
            >
              Sign up
            </button>
          </form>
          <p
            class="text-gray-600 md:hidden text-center text-sm font-light"
          >
            Already have an account?
             
            <a
              class="pl-2 inline-block text-black font-medium"
              href="/auth/signin"
            >
              Log in here
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="md:flex-1 py-5 md:py-0 hidden md:flex items-center justify-center md:h-full bg-gradient-to-br from-teal-600 to-teal-800"
  >
    <div
      class="space-y-6 w-full max-w-xs text-center text-white"
    >
      <h4
        class="text-4xl font-semibold"
      >
        Welcome Back
      </h4>
      <p
        class="font-light text-sm tracking-wide"
      >
        To keep connected with us please login with your personal info
      </p>
      <a
        class="inline-block border border-white rounded-full px-10 py-2"
        href="/auth/signin"
      >
        Sign in
      </a>
    </div>
  </div>
</div>
`;
